File: api/lib/flirtual/user/preferences.ex
defmodule Flirtual.User.Preferences do
  use Flirtual.Schema

  alias Flirtual.User
  alias Flirtual.User.Preferences.{EmailNotifications,Privacy}

  schema "user_preferences" do
    belongs_to :user, User

    has_one :email_notifications, EmailNotifications
    has_one :privacy, Privacy
  end
end

defmodule Flirtual.User.Preferences.EmailNotifications do
  use Ecto.Schema

  alias Flirtual.User.Preferences

  schema "user_preference_email_notifications" do
    belongs_to :preferences, Preferences

    field :matches, :boolean, default: true
    field :messages, :boolean, default: true
    field :likes, :boolean, default: true
    field :newsletter, :boolean, default: true
  end
end

defmodule Flirtual.User.Preferences.Privacy do
  use Ecto.Schema

  alias Flirtual.User.Preferences

  @privacy_enum_values [:everyone, :matches, :me]

  schema "user_preference_privacy" do
    belongs_to :preferences, Preferences

    field :analytics, :boolean, default: true
    field :personality, Ecto.Enum, [values: @privacy_enum_values, default: :everyone]
    field :connections, Ecto.Enum, [values: @privacy_enum_values, default: :everyone]
    field :sexuality, Ecto.Enum, [values: @privacy_enum_values, default: :everyone]
    field :country, Ecto.Enum, [values: @privacy_enum_values, default: :everyone]
    field :kinks, Ecto.Enum, [values: @privacy_enum_values, default: :everyone]
  end
end

File: frontend/src/api/user/preferences.ts
import { fetch, NarrowFetchOptions } from "../exports";

export const PrivacyPreferenceOptions = ["everyone", "matches", "me"] as const;
export type PrivacyPreferenceOption = (typeof PrivacyPreferenceOptions)[number];

export const PreferenceThemes = ["light", "dark", "system"] as const;
export type PreferenceTheme = (typeof PreferenceThemes)[number];

export interface PrivacyPreferences {
	analytics: boolean;
	personality: PrivacyPreferenceOption;
	connections: PrivacyPreferenceOption;
	sexuality: PrivacyPreferenceOption;
	country: PrivacyPreferenceOption;
	kinks: PrivacyPreferenceOption;
}

export interface NotificationPreferences {
	matches: boolean;
	messages: boolean;
	likes: boolean;
	newsletter: boolean;
}

export interface Preferences {
	nsfw: boolean;
	theme: PreferenceTheme;
	emailNotifications: NotificationPreferences;
	privacy: PrivacyPreferences;
}

export async function update(
	userId: string,
	options: NarrowFetchOptions<Partial<Pick<Preferences, "nsfw" | "theme">>>
) {
	return fetch<Preferences>("post", `users/${userId}/preferences`, options);
}

export async function updatePrivacy(
	userId: string,
	options: NarrowFetchOptions<Partial<PrivacyPreferences>>
) {
	return fetch<PrivacyPreferences>("post", `users/${userId}/preferences/privacy`, options);
}

export async function updateNotifications(
	userId: string,
	options: NarrowFetchOptions<Partial<PrivacyPreferences>>
) {
	return fetch<PrivacyPreferences>("post", `users/${userId}/preferences/notifications`, options);
}
